File: Fortran/IBM/execute/F2003/OO_procptr/bindc2/procptrBindcFunptr15.f90
!#######################################################################
!*  ===================================================================
!*  XL Fortran Test Case                          IBM INTERNAL USE ONLY
!*  ===================================================================
!*  ===================================================================
!*
!*  TEST CASE TITLE            :
!*
!*  PROGRAMMER                 : William Zhang 
!*  DATE                       : 3/01/2006
!*  ORIGIN                     : AIX Compiler Development, Toronto Lab
!*                             :
!*
!*  PRIMARY FUNCTIONS TESTED   : Procedure Pointer with BindC 
!*                             :
!*  SECONDARY FUNCTIONS TESTED : 
!*
!*
!*  DESCRIPTION                :  
!*                              pass null function pointer p to C function
!*                              pointer, get address of C function with
!*                              logical as argument and void as return. In
!*                              Fortran, associating p to procedure pointer.
!*                              Referencing procedure pointer entity and
!*                              check the correctness of the return value. 
!* ===================================================================

module fptr15
   use ISO_C_BINDING
   interface
       subroutine csub(cptr) bind(c)
          import C_FUNPTR
          type(C_FUNPTR) :: cptr
       end subroutine csub
   end interface
end module fptr15

program procptrBindcFunptr15

   use ISO_C_BINDING
  
   use fptr15

   interface
       subroutine swap(i, j) bind(c)
          import C_FLOAT_COMPLEX
          complex(C_FLOAT_COMPLEX) :: i, j
       end subroutine 
   end interface

   type(C_FUNPTR) :: p   
   complex(C_FLOAT_COMPLEX) :: i, j

   procedure(csub), pointer :: subind =>null()
   procedure(swap), pointer :: funind => null()

   p = C_NULL_FUNPTR

   i = cmplx(5.0e0,5.0e0) 
   j = cmplx(6.0e0,6.0e0) 

   subind => csub   

   if(C_ASSOCIATED(C_FUNLOC(swap), p)) error stop 1_4
   call subind(p)
   if(.not. C_ASSOCIATED(C_FUNLOC(swap), p)) error stop 2_4

   if(ASSOCIATED(funind)) error stop 3_4
   call C_F_PROCPOINTER(p, funind) 
   if(.not. ASSOCIATED(funind)) error stop 4_4

   if( i .ne. cmplx(5.0e0,5.0e0)) error stop 5_4
   if( j .ne. cmplx(6.0e0,6.0e0)) error stop 6_4

   call funind(i, j)

   if(i .ne. cmplx(6.0e0,6.0e0)) error stop 7_4
   if(j .ne. cmplx(5.0e0,5.0e0)) error stop 8_4

end program procptrBindcFunptr15

File: Fortran/IBM/execute/F2003/OO_procptr/bindc2/procptrBindcFunptr15.c

#include <stdio.h>
#include <stdlib.h>

#ifdef CMPLX
   #include <complex.h>
#else
   #include "cmplx.h"
#endif

void csub(void(**f)(float _Complex *, float _Complex *)) {

   extern void swap(float _Complex *, float _Complex *);

   *f = &swap;

}

void  swap(float _Complex *p1, float _Complex *p2) {

   float _Complex tmp;

#ifdef CMPLX
   if ( *p1 != 5.0f+I*5.0f ) exit(41);
#else
   if ( *p1 != createcomplexf(5.0f,5.0f) ) exit(42);
#endif

#ifdef CMPLX
   if ( *p2 != 6.0f+I*6.0f ) exit(43);
#else
   if ( *p2 != createcomplexf(6.0f,6.0f) ) exit(44);
#endif

   tmp = *p2;
   *p2 = *p1;
   *p1 = tmp;

}

